### PR TITLE
Add rotation gate's matrix representation

### PR DESCRIPTION
**Pull request by Hiroya-W**
## Overview

Added matrix representation of rotated gates to the documentation.

- Tutorials in Qulacs documentaiton
- C++/Python docstring, API documentaiton

`Rot*` gates added by https://github.com/qulacs/qulacs/pull/331.

### CODE REVIEW

**Review by ikanago — APPROVED** [review]
`RotX` series have the common matrix representations which is the same as qiskit and others, right? (This is just my personal question)

**Comment by Hiroya-W** [comment]
> `RotX` series have the common matrix representations which is the same as qiskit and others, right? (This is just my personal question)

Yes. This is often confusing, so the matrix representation was clarified in the document.

File: doc/en/source/guide/2.0_python_advanced.ipynb [hunk 1]
```diff
@@ -1417,7 +1417,44 @@
...
+    "i\\sin(\\frac{\\theta}{2}) & \\cos(\\frac{\\theta}{2})\n",
+    "\\end{pmatrix} \n",
+    "$$ \n",
+    "\n",
+    "`RY`：\n",
+    "$$\n",
+    "R_Y(\\theta) = \\exp(i\\frac{\\theta}{2} Y) = \n",
+    "\\begin{pmatrix} \n",
+    "\\cos(\\frac{\\theta}{2}) & \\sin(\\frac{\\theta}{2}) \\\\ \n",
+    "\\sin(\\frac{\\theta}{2}) & \\cos(\\frac{\\theta}{2})\n",
```
Comment (KowerKoint) [inline doc/en/source/guide/2.0_python_advanced.ipynb:1436]:
```suggestion
    "-\\sin(\\frac{\\theta}{2}) & \\cos(\\frac{\\theta}{2})\n",
```

File: doc/en/source/guide/2.0_python_advanced.ipynb [hunk 2]
```diff
@@ -1417,7 +1417,44 @@
...
+    "\\end{pmatrix} \n",
+    "$$\n",
+    "\n",
+    "`RZ`：\n",
+    "$$\n",
+    "R_Z(\\theta) = \\exp(i\\frac{\\theta}{2} Z) = \n",
+    "\\begin{pmatrix} \n",
+    "e^{i\\frac{\\theta}{2}} & 0 \\\\ \n",
+    "0 & e^{i\\frac{\\theta}{2}}\n",
```
Comment (KowerKoint) [inline doc/en/source/guide/2.0_python_advanced.ipynb:1445]:
```suggestion
    "0 & e^{-i\\frac{\\theta}{2}}\n",
```

**Review by KowerKoint — CHANGES_REQUESTED** [review]
The matrix representations of RY and RZ gate is wrong...

**Comment by Hiroya-W** [comment]
@KowerKoint Thank you very much!!! I didn't check enough.

**Comment by KowerKoint** [comment]
I'm sorry but all of RY/RZ/RotInvY/RotInvZ matrix representations you wrote in this PR are similarly wrong.

File: pysrc/qulacs/__init__.pyi [hunk 3]
```diff
@@ -449,38 +449,137 @@ class QuantumCircuit:
...
     def add_RotX_gate(self, index: int, angle: float) -> None:
         """
         Add Pauli-X rotation gate
+
+        Notes
+        -----
+        Matrix Representation
+
+        .. math::
+            R_X(\\theta) = \exp(-i\\frac{\\theta}{2} X) =
```
Comment (KowerKoint) [inline pysrc/qulacs/__init__.pyi:548]:
```suggestion
            R_X(-\\theta) = \exp(-i\\frac{\\theta}{2} X) =
```

File: pysrc/qulacs/__init__.pyi [hunk 4]
```diff
@@ -449,38 +449,137 @@ class QuantumCircuit:
...
     def add_RotY_gate(self, index: int, angle: float) -> None:
         """
         Add Pauli-Y rotation gate
+
+        Notes
+        -----
+        Matrix Representation
+
+        .. math::
+            R_Y(\\theta) = \exp(-i\\frac{\\theta}{2} Y) =
```
Comment (KowerKoint) [inline pysrc/qulacs/__init__.pyi:563]:
```suggestion
            R_Y(-\\theta) = \exp(-i\\frac{\\theta}{2} Y) =
```

File: pysrc/qulacs/__init__.pyi [hunk 5]
```diff
@@ -449,38 +449,137 @@ class QuantumCircuit:
...
     def add_RotZ_gate(self, index: int, angle: float) -> None:
         """
         Add Pauli-Z rotation gate
+
+        Notes
+        -----
+        Matrix Representation
+
+        .. math::
+            R_Z(\\theta) = \exp(-i\\frac{\\theta}{2} Z) =
```
Comment (KowerKoint) [inline pysrc/qulacs/__init__.pyi:578]:
```suggestion
            R_Z(-\\theta) = \exp(-i\\frac{\\theta}{2} Z) =
```

File: python/cppsim_wrapper.cpp [hunk 6]
```diff
@@ -1213,25 +1213,141 @@ PYBIND11_MODULE(qulacs_core, m) {
...
+            py::arg("index"), py::arg("angle"))
+        .def("add_RotX_gate", &QuantumCircuit::add_RotX_gate, R"(
+                Add Pauli-X rotation gate
+
+                Notes
+                -----
+                Matrix Representation
+
+                .. math::
+                    R_X(\\theta) = \exp(-i\\frac{\\theta}{2} X) =
```
Comment (KowerKoint) [inline python/cppsim_wrapper.cpp:1314]:
```suggestion
                    R_X(-\\theta) = \exp(-i\\frac{\\theta}{2} X) =
```

File: python/cppsim_wrapper.cpp [hunk 7]
```diff
@@ -1213,25 +1213,141 @@ PYBIND11_MODULE(qulacs_core, m) {
...
+            py::arg("index"), py::arg("angle"))
+        .def("add_RotY_gate", &QuantumCircuit::add_RotY_gate, R"(
+                Add Pauli-Y rotation gate
+
+                Notes
+                -----
+                Matrix Representation
+
+                .. math::
+                    R_Y(\\theta) = \exp(-i\\frac{\\theta}{2} Y) =
```
Comment (KowerKoint) [inline python/cppsim_wrapper.cpp:1329]:
```suggestion
                    R_Y(-\\theta) = \exp(-i\\frac{\\theta}{2} Y) =
```

File: python/cppsim_wrapper.cpp [hunk 8]
```diff
@@ -1213,25 +1213,141 @@ PYBIND11_MODULE(qulacs_core, m) {
...
+            py::arg("index"), py::arg("angle"))
+        .def("add_RotZ_gate", &QuantumCircuit::add_RotZ_gate, R"(
+                Add Pauli-Z rotation gate
+
+                Notes
+                -----
+                Matrix Representation
+
+                .. math::
+                    R_Z(\\theta) = \exp(-i\\frac{\\theta}{2} Z) =
```
Comment (KowerKoint) [inline python/cppsim_wrapper.cpp:1344]:
```suggestion
                    R_Z(-\\theta) = \exp(-i\\frac{\\theta}{2} Z) =
```

File: src/cppsim/circuit.hpp [hunk 9]
```diff
@@ -410,7 +473,17 @@ class DllExport QuantumCircuit {
     /**
      * \~japanese-en X-rotation gateを追加する。
      *
-     * ゲートの表記は \f$ R_X(\theta) = \exp(-i\theta X) \f$になっている。
+     * ゲートの表記は
+     *
+     * @f[
+     * R_X(\theta) = \exp(-i\frac{\theta}{2} X) =
```
Comment (KowerKoint) [inline src/cppsim/circuit.hpp:479]:
```suggestion
     * R_X(-\theta) = \exp(-i\frac{\theta}{2} X) =
```

File: src/cppsim/circuit.hpp [hunk 10]
```diff
@@ -420,7 +493,17 @@ class DllExport QuantumCircuit {
     /**
      * \~japanese-en Y-rotation gateを追加する。
      *
-     * ゲートの表記は \f$ R_Y(\theta) = \exp(-i\theta Y) \f$になっている。
+     * ゲートの表記は
+     *
+     * @f[
+     * R_Y(\theta) = \exp(-i\frac{\theta}{2} Y) =
```
Comment (KowerKoint) [inline src/cppsim/circuit.hpp:499]:
```suggestion
     * R_Y(-\theta) = \exp(-i\frac{\theta}{2} Y) =
```

File: src/cppsim/circuit.hpp [hunk 11]
```diff
@@ -430,7 +513,18 @@ class DllExport QuantumCircuit {
     /**
      * \~japanese-en Z-rotation gateを追加する。
      *
-     * ゲートの表記は \f$ R_Z(\theta) = \exp(-i\theta Z) \f$になっている。
+     * ゲートの表記は
+     *
+     * @f[
+     * R_Z(\theta) = \exp(-i\frac{\theta}{2} Z) =
```
Comment (KowerKoint) [inline src/cppsim/circuit.hpp:519]:
```suggestion
     * R_Z(-\theta) = \exp(-i\frac{\theta}{2} Z) =
```

File: src/cppsim/gate_factory.hpp [hunk 12]
```diff
@@ -226,6 +236,16 @@ DllExport ClsOneQubitRotationGate* RX(UINT qubit_index, double angle);
 /**
  * \~japanese-en \f$Y\f$回転ゲートを作成する。
  *
+ * @par Matrix Representation
+ *
+ * @f[
+ * R_Y(\theta) = \exp(i\frac{\theta}{2} Y) =
+ *     \begin{pmatrix}
+ *     \cos(\frac{\theta}{2})  & \sin(\frac{\theta}{2}) \\
+ *     \sin(\frac{\theta}{2}) & \cos(\frac{\theta}{2})
```
Comment (KowerKoint) [inline src/cppsim/gate_factory.hpp:245]:
```suggestion
 *     -\sin(\frac{\theta}{2}) & \cos(\frac{\theta}{2})
```

File: src/cppsim/gate_factory.hpp [hunk 13]
```diff
@@ -235,6 +255,16 @@ DllExport ClsOneQubitRotationGate* RY(UINT qubit_index, double angle);
 /**
  * \~japanese-en \f$Z\f$回転ゲートを作成する。
  *
+ * @par Matrix Representation
+ *
+ * @f[
+ * R_Z(\theta) = \exp(i\frac{\theta}{2} Z) =
+ *     \begin{pmatrix}
+ *     e^{i\frac{\theta}{2}} & 0 \\
+ *     0 & e^{i\frac{\theta}{2}}
```
Comment (KowerKoint) [inline src/cppsim/gate_factory.hpp:264]:
```suggestion
 *     0 & e^{-i\frac{\theta}{2}}
```

**Review by KowerKoint — CHANGES_REQUESTED** [review]

**Review by KowerKoint — APPROVED** [review]

**Comment by KowerKoint** [comment]
I fixed some more documents, but I'm insecure to merge it main immediately.
@ikanago Could you re-review this?

File: pysrc/qulacs/__init__.pyi [hunk 14]
```diff
@@ -449,38 +449,137 @@ class QuantumCircuit:
...
+        Notes
+        -----
+        Matrix Representation
+
+        .. math::
+            R_X(-\\theta) = \exp(-i\\frac{\\theta}{2} X) =
+                \\begin{pmatrix}
+                \cos(\\frac{\\theta}{2})  & -i\sin(\\frac{\\theta}{2}) \\\\
+                -i\sin(\\frac{\\theta}{2}) & \cos(\\frac{\\theta}{2})
+                \end{pmatrix}
```
Comment (ikanago) [inline pysrc/qulacs/__init__.pyi:552]: Matrix representation of `RotX` should be like
```suggestion
        .. math::
            RotX(\\theta) = \exp(-i\\frac{\\theta}{2} X) =
                \\begin{pmatrix}
                \cos(\\frac{\\theta}{2})  & -i\sin(\\frac{\\theta}{2}) \\\\
                -i\sin(\\frac{\\theta}{2}) & \cos(\\frac{\\theta}{2})
                \end{pmatrix}
```

- `RotX`, not `R_X`
- Negation in the argument is not needed

**Review by ikanago — CHANGES_REQUESTED** [review]

File: pysrc/qulacs/__init__.pyi (hunk 14 above)
Comment (KowerKoint) [reply pysrc/qulacs/__init__.pyi]: changed to `RotX`

**Review by ikanago — APPROVED** [review]
Thank you for fix!  LGTM 👍

**Pull request merged**